README.md
# H5边滑边动模板

![version](https://img.shields.io/github/package-json/v/zengxue158/h5-template-scroller.svg)
![commit](https://img.shields.io/github/last-commit/zengxue158/h5-template-scroller.svg)
![new feature](https://img.shields.io/badge/author-zeng_xue-orange)

<!-- ![code-size](https://img.shields.io/github/languages/code-size/NyPhile/h5_template.svg) -->
<!-- ![lang](https://img.shields.io/github/languages/top/NyPhile/h5_template.svg) -->

### 安装

```bash
$ npx ne-build init 项目名 -t zengxue158/h5-template-scroller
# node版本大于5.2
# 参数可在命令行输入，也可以输入命令后按交互提示输入

# 参数⬇️
-n or --name
# 项目名称，必须
# 用于替换模板项目中package.json的{name}

-c or --channel
# 频道名称，必须
# 用于替换模板项目内上传路径中的频道路径

-d or --desc
# 项目描述，可选
# 用于替换模板项目中package.json的{description}

--username
# 上传工具账号，必须，即邮箱前缀

--password
# 上传工具密码，必须，即邮箱密码
```

### 命令

```bash
# 安装依赖
$ npm i

# 开启本地服务
$ npm run dev

# 打包（上传前需先打包）
$ npm run build

# 线上环境测试
# 发布到 https://wp.m.163.com/163/test/{channel}/{name}/index.html
$ gulp test

# 发布
$ gulp publish
# 发布到 https://wp.m.163.com/163/page/{channel}/{name}/index.html

# 清空本地缓存文件，使所有文件重新上传
# （默认每次上传只上传增量文件）
$ gulp clear
```

### 目录结构

```
h5_template
 ├ src/
 │  ├ assets/               # 图片等资源文件夹
 │  ├ css/
 │  │  ├ common.less        # 公用reset css
 │  │  └ index.less         # 样式文件
 │  ├ js/
 │  │  ├ utils/             # 工具js文件夹
 │  │  │  ├ helper.js       # eruda、统计js
 │  │  │  ├ share.js        # 分享
 │  │  │  ├ track.js        # 统计方法
 │  │  │  ├ utils.js        # 工具函数
 │  │  │  └ viewport.js     # 屏幕适配
 │  │  ├ Animate.js         # 动画库
 │  │  ├ common.js          # 通用js
 │  │  ├ hilo-standalone.js # 游戏引擎
 │  │  ├ imglist.js         # 交互动画参数
 │  │  ├ Scroller.js        # 滚动插件
 │  │  └ index.js           # js文件
 ├ static/                  # 静态资源文件夹，会直接上传至 https://static.ws.126.net/163/f2e/{channel}/{name}/static/ 下
 ├ .eslintrc.js             # eslint配置文件
 ├ .ftppass                 # 上传账号密码文件
 ├ .gitignore
 ├ babel.config.js          # babel配置文件
 ├ gulpfile.js              # gulp配置文件
 ├ package.json
 ├ postcss.config.js        # postcss配置文件
 ├ README.md
 ├ webpack.base.config.js   # webpack通用设置
 ├ webpack.dev.config.js    # webpack开发设置
 └ webpack.prod.config.js   # webpack发布设置
```

### 简介

* 项目使用webpack打包，webpack-dev-server开启本地服务
* html默认使用.html，可按照[html-webpack-plugin文档](https://github.com/jantimon/html-webpack-plugin)根据需求自行配置模板文件。
* css默认使用[less](http://lesscss.org/)以及[postcss](https://postcss.org/)
* js使用[babel7.5](https://babeljs.io/)进行编译，按照package.json中的browserslist自动增加polyfill
* eslint使用默认标准，[规则](https://eslint.org/docs/rules/)
* 统计ID位于`package.json`的`projectId`字段
* 已内置`@mf2e/js-bridge`，用于H5调用客户端方法，[项目](http://npm.hz.netease.com/package/@mf2e/js-bridge)，[文档](https://wp.m.163.com/163/html/newsclient/api/index.html#/)
* 已内置`@newsapp-activity/newsapp-share`，`src/js/utils/share.js`对其进行封装
* 已内置`Zepto`，版本`v1.2.0`，包含模块`zepto`、`event`、`ajax`、`form`、`ie`、`detect`、`fx`、`fx_methods`、`assets`、`data`、`deferred`、`callbacks`、`selector`、`touch`、`gesture`、`stack`、`ios3`，[文档](https://zeptojs.com/)。（由于Zepto非模块化js，不能直接打包引入，因此在index.html中引入js，webpack.externals进行排除）
* npm run dev开启本地服务后，可以使用localhost:8080或{ip}:8080或dev.f2e.163.com:8080（需配置hosts）调试，可使用`eruda`（[文档](https://eruda.liriliri.io/)）进行移动端调试
* 打包时静态资源默认添加hash，上传后进行缓存，无修改不再重复上传

### 说明

1. 所有图片动画在 `src/js/imglist.js` 内
2. 动画格式⬇️。精灵动画与transitions动画可叠加，animations与transitions动画可叠加；精灵动画与animations由于play参数公用，不能叠加，只执行精灵动画

```js
    // 基本属性
    {
      id: 'btn_more',   // 元素id，必须，不能重复，用于存储Hilo.View对象以及loadQueue中的id
      mode: 'Bitmap',   // Hilo.View 类型，必须
      parent: 'p5',     // 父级元素
      src: btn_more,    // 图片链接，import进来
      delay: delay.p5,  // 整体模块delay的时间
      mask: 'cover_mask'// mask遮罩
      propes: {x: 70, y: 4370, origin: '50% 50%'}, // Hilo.View的属性。其中增加origin，类似transform-origin，使用百分比控制中心点，pivotX、x自动计算，例如：loading_gear_big
      draw () {},       // mode=Graphics时必须，用于画Graphics
      touchend () {},    // 点击事件，this指向app，用touchstart与touchend间的位移距离排除误触
      transitions: [{propes: 'y', time: [0, 0 + 200], value: [0 + 150, 0]}], // 位移动画，ease默认线性
    }
    // mask遮罩，先建一个Graphics，其他元素的mask属性指向这个mask的id，即可只展示该mask区域内的部分，例如：cover_mask、cover_bg
    // 几种动画写法
    // 1.平移动画，一开始就播放，例如：loading_gear_big
    {
      animations: [
        [{rotation: 30}, {time: 5000, delay: 500, ease: Hilo.Ease.Quad.EaseOut}]
      ]
    }
    // 2.平移动画，滚动到某处播放，无例子
    {
      play: 1000, // scroller.scrollTop值
      animations: [
        [{rotation: 30}, {time: 5000, delay: 500, ease: Hilo.Ease.Quad.EaseOut}]
      ]
    }
    // 3.平移动画，与滚动位置绑定，例如：p1_qout
    {
      transitions: [
        {propes: 'y', time: [0, 5000], value: [winHeight, winHeight - 5000], ease: Hilo.Ease.Linear.easeNoneFn}
      ]
    }
    // 4.Sprite精灵动画，一开始就播放，例如：cover_btn
    {
      mode: 'Sprite',
      atlas: {
        width: 1804,        // sprite图片宽度
        height: 4104,       // sprite图片高度
        frames: {
          frameWidth: 451,  // 帧宽度
          frameHeight: 684, // 帧高度
          numFrames: 24     // 帧总数
        },
        sprites: {          // 若播放帧数是全部帧(如一共10帧，from 0 to 9)，可去掉此对象，例如：p1_smoke
          sprite: {         // Hilo.TextureAtlas 序列id
            from: 0,        // 起始帧序号
            to: 23          // 终止帧序号
          }
        }
      },
      frames: 'sprite'      // Hilo.Sprite 取对应的 Hilo.TextureAtlas 的序列id；若播放帧数是全部帧，可去掉此对象
    }
    // 5.Sprite精灵动画，到某处播放，例如：p1_text1_in
    // 在4.基础上增加
    {
      play: 1000  // 播放时间
    }
    // 6.Sprite精灵动画，与距离绑定，例如：p1_bg_out
    // 在4.基础上增加
    {
      play: {start: 4200, end: 4200 + 200} // 播放起止位置，自动计算起始~终止间播放帧
    }

  ```


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.0.0] - 2019-07-24
### Added
- 基础版本

## [1.0.1] - 2019-10-09
### Added
- tinify-loader 增加图片压缩处理

### Changed
- 修改了devServer的热更新模式

## [1.0.2] - 2019-10-18
### Fixed
- 升级eslint，修复`eslint-utils@1.4.0`的安全漏洞

### Changed
- 还原了devServer的热更新模式。

## 2019-11-20
### Changed
- 将`manifest.js`改为inline调用
- 将`main.js`改为async调用
- 增加了章鱼以及回流统计。章鱼统计发布到章鱼系统`mapp`站点；回流统计发布到章鱼系统的`sps`站点，modelid为{频道名} + {项目名}。



src/js/index.js
import '../css/index.less' // 引入css
import './common.js' // 公共js
import { shareWithConfig } from 'utils/share.js'
import './hilo-standalone.js'
import Scroller from './Scroller.js'
import { isArray } from 'utils/utils.js'
import { getImgList } from './imglist.js'  //所有动画参数写在这里

var isWeixin = navigator.userAgent.match(/micromessenger/i);

let app = {
  canvas: $('#stage'),
  stage: null,
  winWidth: 0,
  winHeight: 0,
  imgList: [],
  stageContent: {},
  loadQueue: {},
  musics:[ // 滑动到固定位置播放音频、停止音频
    // {id: "", start: ,end:},
  ],
  init() {
    let self = this
    this.loadQueue = this.getLoadQueue(this.start.bind(this))
    // 背景音乐
    // this.autoplay('bgm')
  },
  bind() {
    let self = this
    // 背景音乐兼容性处理
    // $('body').one('touchend', () => {
    //   $('#bgm')[0].play()
    //   if(!isWeixin){ //为了兼容其他客户端边滑边动播放音乐的问题。
    //     for(var i=0;i<app.musics.length;i++){
    //       $("#"+app.musics[i].id)[0].play()
    //     }
    //   }
    // })
    // 阻止页面上下弹动
    // document.querySelector(".loading").addEventListener("touchmove",function(e){e.stopPropagation();e.preventDefault()},false);
    
  },
  autoplay(id) {
    let audioAutoPlay = () => {
      const audio = $('#' + id)[0]
      audio.play().catch((e) => {
        console.log(e)
      })
      document.addEventListener(
        'WeixinJSBridgeReady',
        () => {
          audio.play()
        },
        false
      )
    }
    audioAutoPlay()
  },
  initMusic(){
    let self = this;
    if(app.musics.length>0){
      for (var i = 0; i < app.musics.length; i++) {
        (function(i){
          var audio = $("#" + app.musics[i].id)[0];
          var play = function() {
            document.removeEventListener("WeixinJSBridgeReady", play);
            document.removeEventListener("YixinJSBridgeReady", play);
            audio.play();
          };
          $(audio).one("play", function(){
            this.pause();
          })
          document.addEventListener("WeixinJSBridgeReady", play, false);
          document.addEventListener('YixinJSBridgeReady', play, false);
        })(i);
      }
    }
  },
  getLoadQueue(completeCb) {
    let percent = $('#loading_num')
    this.imgList = getImgList()
    let imgList = this.imgList.filter((item) => item.preload !== false && item.src)

    let queue = new Hilo.LoadQueue()
    queue.add(imgList)

    queue.on('load', (e) => {
      percent.text(`${(e.target._loaded / e.target._source.length * 100).toFixed()}%`)
    })

    queue.on('complete', () => {
      completeCb()
      queue.off('complete').off('load');
      
    })

    return queue.start()
  },
  start() {
    // loading结束后
    this.initStage()
    this.addViews()
    this.initScroller()

  },
  initStage() {
    ;(this.winWidth = $('body').width()), (this.winHeight = $('body').height() > 1448 ? 1448 : $('body').height())
    this.canvas.attr({ width: this.winWidth, height: this.winHeight })
    this.stage = new Hilo.Stage({
      canvas: this.canvas[0],
      width: this.winWidth,
      height: this.winHeight
    })
    let ticker = new Hilo.Ticker(60)
    ticker.addTick(Hilo.Tween)
    ticker.addTick(this.stage)
    ticker.start()
    window.stage = stage
    this.stage.enableDOMEvent(Hilo.event.POINTER_START, true)
    this.stage.enableDOMEvent(Hilo.event.POINTER_MOVE, true)
    this.stage.enableDOMEvent(Hilo.event.POINTER_END, true)
  },
  renderStage(resources) {
    let self = this
    if (!isArray(resources)) {
      resources = [ resources ]
    }
    resources.forEach((item) => {
      renderResource(item)
    })
    function renderResource(resource) {
      resource.propes = resource.propes || {}
      // parent 非atlas放在stage下，其他置于对应parent
      let parent =
				!resource.parent || resource.parent == 'stage' ? self.stage : self.stageContent[resource.parent]
      // 遮罩
      if (resource.mask) {
        resource.propes.mask = self.stageContent[resource.mask]
      }
      // y值居中计算时使用center
      if (resource.propes && resource.propes.center) {
        resource.propes.y = self.winHeight / 2 + resource.propes.center
      }
      // y值底部对齐使用bottom
      if (resource.propes && resource.propes.bottom) {
        resource.propes.y = self.winHeight - resource.propes.bottom
      }
      // scaleX == scaleY时可使用scale
      if (resource.propes && resource.propes.scale) {
        resource.propes.scaleX = resource.propes.scaleY = resource.propes.scale
      }
      // 获取图片
      if (resource.src) {
        let imageId = resource.imageId || resource.id
        if (resource.mode == 'Sprite') {
          resource.atlas.image = self.loadQueue.get(imageId).content
        } else {
          resource.propes.image = self.loadQueue.get(imageId).content
        }
      }
      // new Hilo.TextureAtlas
      if (resource.mode == 'Sprite' && resource.atlas) {
        if (!resource.atlas.sprites && resource.atlas.frames && resource.atlas.frames.numFrames) {
          let lastFrame = resource.atlas.frames.numFrames - 1
          resource.atlas.sprites = { sprite: { from: 0, to: lastFrame } }
          resource.frames = 'sprite'
        }
        self.stageContent[resource.id + '_atlas'] = new Hilo.TextureAtlas(resource.atlas)
      }
      // frames
      if (resource.frames) {
        if (resource.frames.indexOf('.') > 0) {
          let [ framesId, framesName ] = resource.frames.split('.')
          resource.propes.frames = self.stageContent[framesId].getSprite(framesName)
        } else {
          resource.propes.frames = self.stageContent[resource.id + '_atlas'].getSprite(resource.frames)
        }
      }
      // new Hilo.View
      let item = (self.stageContent[resource.id] = new Hilo[resource.mode](resource.propes))
      item.addTo && item.addTo(parent)
      // 类似css的transform-origin百分比，'50% 50%'，坐标值会自动重新计算
      if (resource.propes && resource.propes.origin) {
        let [ originX, originY ] = resource.propes.origin.split(' ')
        let percentX = parseFloat(originX) / 100
        let percentY = parseFloat(originY) / 100
        item.pivotX = percentX * item.width
        item.pivotY = percentY * item.height
        item.x += item.pivotX
        item.y += item.pivotY
      }
      // 绘制矢量图
      if (resource.mode == 'Graphics') {
        resource.draw.call(item)
      }
      // 自动播放动效
      if (resource.animations && !resource.play) {
        resource.animations.forEach((animation) => {
          let tween = new Hilo.Tween.to(item, animation[0], animation[1])
          window.tween = window.tween || tween
        })
      }
      // 点击事件 排除滑动后触发
      if (resource.touchend) {
        item.on('touchstart', (e) => {
          self.touchX = e.touches[0].pageX
          self.touchY = e.touches[0].pageY
        })
        item.on('touchend', (e) => {
          if (
            Math.abs(e.changedTouches[0].pageX - self.touchX) < 10 &&
						Math.abs(e.changedTouches[0].pageY - self.touchY) < 10
          ) {
            resource.touchend.call(self)
          }
        })
      }
    }
    window.stageContent = this.stageContent
  },
  addViews() {
    this.renderStage(this.imgList)
  },
  render(l, t, z) {
    let self = this
    // console.log(t)
    this.imgList.forEach((resource) => {
      let delay = resource.delay || 0,
        item = this.stageContent[resource.id]

      if (resource.play && resource.play.start && resource.mode == 'Sprite') {
        let times = resource.play,
          during = times.end - times.start,
          length = item.getNumFrames(),
          step = during / length / (resource.playCount ? resource.playCount : 1),
          time_start = times.start - 400,
          time_end = times.end + 400
        if (t > time_start + delay && t < times.start + delay) {
          item.goto(0, true)
        } else if (t < time_end + delay && t > times.end + delay) {
          item.goto(length - 1, true)
        } else if (t >= resource.play.start + delay && t <= resource.play.end + delay) {
          let frameIndex = parseInt((t - resource.play.start - delay) / step)
          frameIndex = frameIndex >= length ? frameIndex - length : frameIndex
          item.goto(frameIndex, true)
        }
      } else if (resource.play && resource.mode == 'Sprite') {
        // sprite 固定位置动画
        if (t > resource.play + delay && item.paused) {
          item.play()
        } else if (t < resource.play + delay - 400 && item.currentFrame != 0) {
          item.goto(0, true)
        }
      } else if (resource.play && resource.animations) {
        // animation 固定位置动画
        resource.tweens = resource.tweens || []
        if (t > resource.play + delay) {
          resource.animations.forEach((animation, index) => {
            if (!resource.tweens[index]) {
              animation[1].paused = true
              resource.tweens[index] = new Hilo.Tween.to(item, animation[0], animation[1])
              resource.tweens[index].start()
            }
          })
        } else if (t < resource.play + delay - 400) {
          resource.animations.forEach((animation, index) => {
            if (resource.tweens[index]) {
              for (let prop in resource.tweens[index]._fromProps) {
                item[prop] = resource.tweens[index]._fromProps[prop]
              }
              Hilo.Tween.remove(resource.tweens[index])
              resource.tweens[index] = null
            }
          })
        }
      }
      // transition 边划变动动画
      let transitions = resource.transitions
      if (!transitions) return
      transitions.forEach((transition) => {
        let values = transition.value,
          times = transition.time,
          time_start = times[0] - 400,
          time_end = times[times.length - 1] + 400
        if (t >= time_start + delay && t <= times[0] + delay) {
          item[transition.propes] = values[0]
        } else if (t < time_end + delay && t > times[times.length - 1] + delay) {
          item[transition.propes] = values[values.length - 1]
        } else {
          values.forEach((value, index) => {
            if (t < times[index + 1] + delay && t > times[index] + delay) {
              let ease = transition.ease || Hilo.Ease.Linear.EaseNone
              let k = (t - times[index] - delay) / (times[index + 1] - times[index])
              item[transition.propes] = values[index] + ease(k) * (values[index + 1] - values[index])
            }
          })
        }
      })
    })
    // 滑动到固定位置播放音乐 需用到的 
    // for (var l = 0; l < app.musics.length; l++) {
    //   var music = app.musics[l];
    //   if(music.start && t < music.start){
    //     music.played = "";
    //     // el.attr("data-played", "")
    //     $("#"+music.id)[0].pause(); 
    //   }else if(music.end && t >= music.end){
    //     $("#"+music.id)[0].pause();
    //     music.played = "";
    //     // el.attr("data-played", "")
    //   }else if(music.start && t >= music.start && !music.played){
    //     $("#"+music.id)[0].play(); 
    //     music.played = true;
    //   }
    // }
  },
  shareShow() {
    shareWithConfig();
  },
  againFun(){
    function updateUrl(url,key){
      var key= (key || 't') +'=';  //默认是"t"
      var reg=new RegExp(key+'\\d+');  //正则：t=1472286066028
      var timestamp=+new Date();
      if(url.indexOf(key)>-1){ //有时间戳，直接更新
        return url.replace(reg,key+timestamp);
      }else{  //没有时间戳，加上时间戳
        if(url.indexOf('\?')>-1){
          var urlArr=url.split('\?');
          if(urlArr[1]){
            return urlArr[0]+'?'+key+timestamp+'&'+urlArr[1];
          }else{
            return urlArr[0]+'?'+key+timestamp;
          }
        }else{
          if(url.indexOf('#')>-1){
            return url.split('#')[0]+'?'+key+timestamp+location.hash;
          }else{
            return url+'?'+key+timestamp;
          }
        }
      }
    }
    window.location.href=updateUrl(window.location.href); //不传参，默认是“t”
  },
  initScroller() {
    this.canvas.on('touchstart', this.drawStart).on('touchmove', this.drawMove).on('touchend', this.drawEnd)
    this.scroller = new Scroller(this.render.bind(this), {
      // bouncing: false,
      animationDuration: 1000,
      speedMultiplier:2.5  //乘以或减少滚动速度
    })
    // 228120 这是整体滚动距离 需要修改
    this.scroller.setDimensions(this.winWidth, this.winHeight, this.winWidth, 228120 + this.winHeight)
    window.scroller = this.scroller
  },
  drawStart(event) {
    event.stopPropagation(), event.preventDefault()
    app.isTouched = true
    app.scroller.doTouchStart(event.touches, event.timeStamp)
  },
  drawMove(event) {
    if (!app.isTouched) return

    event.stopPropagation(), event.preventDefault()
    app.isTouched = true
    app.scroller.doTouchMove(event.touches, event.timeStamp, event.scale)
  },
  drawEnd(event) {
    app.scroller.doTouchEnd(event.timeStamp)
    app.isTouched = false
  },
  /*
    @param width 图片宽度
    @param height 图片高度
    @param column sprite 横向个数
    @param number 总数
   */
  getFrames(width, height, column, number) {
    let arr = [],
      x = 0,
      y = 0
    for (let i = 0; i < number; i++) {
      let frame = [ x * width, y * height, width, height ]
      arr.push(frame)
      if (x + 1 < column) {
        x = x + 1
      } else {
        x = 0
        y = y + 1
      }
    }
    return arr
  }
}
$.fn.ready(() => {
  // 播放背景音乐、滑动到固定位置播放音乐
  // app.autoplay('bgm');
  // app.initMusic();
})

setTimeout(() => {
  app.init();
  app.bind();
}, 700)